### ISSUE TITLE
Retrieve sub clusters of one specific cluster

### ISSUE THREAD

**Issue by Karol-G** [scikit-learn-contrib/hdbscan#442]
Hi,

I want to get the sub clusters of one specific cluster. I know there is the `get_clusters` function but it returns simply all clusters of another level, so not really what I want. Isn't there some function that expects a cluster label as input and returns for all points within that cluster new cluster labels? Essentially the same as calling `fit` on the points of this cluster again.

Best
Karol

**Comment by Karol-G** [comment]
I tried to solve this but it does not seem to work quite yet. Maybe my approach is wrong :/
This is what I am doing:

1. Generate blob data
2. Cluster all data
3. Filter the linkage by removing all links that don't belong to my cluster label of interest
4. Convert this filtered linkage to flat clusters with `hdbscan.hdbscan_._tree_to_labels` and return the new cluster labels
5. Filter the original data so that only the data of my cluster label of interest remain
6. Plot 

However the plot shows that the new sub clusters are intermixed with each other. So either I simply have a bug somewhere or the approach in general is not possible and that you can't call `_tree_to_labels` on a modified linkage. 
Can you call `_tree_to_labels` on a modified linkage or will the result be wrong?

Here is my code:

``` python
import hdbscan
import numpy as np
from sklearn.datasets import make_blobs
import seaborn as sns
import matplotlib.pyplot as plt

def filter_linkage(linkage, labels, label):
    length = len(labels)
    indices = np.nonzero([labels == label])[1].tolist()
    indices_length = len(indices)
    filtered_linkage = []
    new_index_counter = 0
    new_cluster_counter = indices_length
    original_indices = []
    for i, link in enumerate(linkage):
        if link[0] in indices and link[1] in indices:
            new_link = []
            for j in [0, 1]:
                if link[j] < length:
                    new_link.append(new_index_counter)
                    original_indices.append([int(link[j]), new_index_counter])
                    new_index_counter += 1
                else:
                    new_link.append(new_cluster_counter)
                    original_indices.append([int(link[j]), new_cluster_counter])
                    new_cluster_counter += 1
            new_link.append(link[2])
            new_link.append(link[3])
            filtered_linkage.append(new_link)
            indices.append(i+length)
    filtered_linkage = np.asarray(filtered_linkage)
    original_indices = np.asarray(original_indices)
    return filtered_linkage, original_indices, indices_length

def filter_data(data, original_indices, indices_length):
    filtered_data = [None]*indices_length
    for index in original_indices:
        if index[1] < indices_length:
            filtered_data[index[1]] = data[index[0]]
    filtered_data = np.asarray(filtered_data)
    return filtered_data


# Generate blob data
data, _ = make_blobs(1000)

# Cluster all data
clusterer = hdbscan.HDBSCAN(min_cluster_size=10)
cluster_labels = clusterer.fit_predict(data)

# Filter linkage by removing all links that don't belong to cluster_label
cluster_label = 1
new_linkage, original_indices, indices_length = filter_linkage(clusterer._single_linkage_tree, cluster_labels, cluster_label)

# Convert linkage to flat clusters and return the new cluster labels
new_labels = hdbscan.hdbscan_._tree_to_labels(0, single_linkage_tree=new_linkage, min_cluster_size=10)[0]
print(new_labels)

# Filter the original data so that only the data of cluster_label remains
new_data = filter_data(data, original_indices, indices_length)

# Plot
color_palette = sns.color_palette('deep', 8)
cluster_colors = [color_palette[x] if x >= 0
                  else (0.5, 0.5, 0.5)
                  for x in new_labels]
cluster_member_colors = [sns.desaturate(x, p) for x, p in
                         zip(cluster_colors, clusterer.probabilities_)]
plt.scatter(*new_data.T, s=50, linewidth=0, c=cluster_member_colors, alpha=0.25)
plt.show()
```

Here is an example plot that shows the problem of intermixed sub clusters:
![grafik](https://user-images.githubusercontent.com/3471895/103406712-f6d20c00-4b5b-11eb-8fd5-ea4b701fba17.png)


I hope someone can help me with this :)

Best
Karol
